db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java

```diff
@@ -1601,39 +1599,37 @@ public ResultSetNode optimize(DataDictionary dataDictionary,
                 }
             }
         }
-		/* Get a new optimizer */
-        optimizer=getOptimizer(fromList, wherePredicates, dataDictionary, orderByList);
-        optimizer.setOuterRows(outerRows);
 
-        // Aggregation with no GROUP BY always outputs one row.
-        if ((selectAggregates != null && !selectAggregates.isEmpty() ||
-             havingAggregates !=null  && !havingAggregates.isEmpty()) && hasNoGroupBy())
-            optimizer.setSingleRow(true);
+        Optimizer optimizer = getOptimizer(fromList, wherePredicates, dataDictionary, orderByList);
+        if (nestingLevel == 0) {
+            // Top level select node, we should try to cost control first, and fallback to spark if necessary
 
-		/* Optimize this SelectNode */
-        while(optimizer.nextJoinOrder()){
-            while(optimizer.getNextDecoratedPermutation()){
-                optimizer.costPermutation();
+            if (!determineSpark()) {
+                optimizer.setForSpark(false);
+                findBestPlan(optimizer, dataDictionary, outerRows);
+            }
+
+            if (determineSpark()) {
+                for (int i = 0; i < fromList.size(); ++i) {
+                    FromTable ft = (FromTable) fromList.elementAt(i);
+                    ft.resetAccessPaths();
+                }
+                optimizer.setForSpark(true);
+                optimizer.prepForNextRound();
+                findBestPlan(optimizer, dataDictionary, outerRows);
             }
+        } else {
+            findBestPlan(optimizer, dataDictionary, outerRows);
```

Comment:
We use different OptimizerImpl instance to plan each SELECT block. We set the useSpark variable for the OptimizerImpl instance for the top SELECT, but I don't see logic that passes the setting to the OptimizerImpl instance for the nested SELECT block.
Given the following query as an example:
```
explain select * from (select * from t1 --splice-properties useDefaultRowCount=30000000
, t2 where a1=a2) dt1, (select * from t3, t4 where a3=a4) dt2 where a1=a2; 
```
When the top SELECT is planning for the spark path, the nested SELECT still has useSpark=false.

Reply:
Done